File: tokyo-x/src/modules/jobs/task-manager.ts
import { appendFileSync, mkdirSync } from "node:fs";
import { join } from "node:path";
import { randomUUID } from "node:crypto";

export type JobStatus = "queued" | "running" | "done" | "failed" | "cancelled";

export interface Job {
  id: string;
  kind: string;
  payload: unknown;
  status: JobStatus;
  progress: number;
  result?: unknown;
  error?: string;
  createdAt: string;
  startedAt?: string;
  finishedAt?: string;
}

export type Runner = (payload: unknown, job: Job, report: (progress: number) => void) => Promise<unknown>;

export class TaskManager {
  private jobs = new Map<string, Job>();
  private queue: string[] = [];
  private runners = new Map<string, Runner>();
  private active = 0;
  private pumpTimer: ReturnType<typeof setInterval>;

  constructor(private readonly eventDir: string, private readonly concurrency = 2) {
    mkdirSync(eventDir, { recursive: true });
    this.pumpTimer = setInterval(() => this.pump(), 250);
    this.pumpTimer.unref?.();
  }

  register(kind: string, runner: Runner): void {
    this.runners.set(kind, runner);
  }

  submit(kind: string, payload: unknown): Job {
    if (!this.runners.has(kind)) throw new Error(`no runner registered for kind "${kind}"`);
    const job: Job = {
      id: `job_${randomUUID().slice(0, 10)}`,
      kind,
      payload,
      status: "queued",
      progress: 0,
      createdAt: new Date().toISOString(),
    };
    this.jobs.set(job.id, job);
    this.queue.push(job.id);
    this.event(job, "queued");
    setImmediate(() => this.pump());
    return { ...job };
  }

  cancel(id: string): boolean {
    const job = this.jobs.get(id);
    if (!job) return false;
    if (job.status === "queued") {
      this.queue = this.queue.filter((x) => x !== id);
      job.status = "cancelled";
      job.finishedAt = new Date().toISOString();
      this.event(job, "cancelled");
      return true;
    }
    return false;
  }

  get(id: string): Job | undefined {
    const j = this.jobs.get(id);
    return j ? { ...j } : undefined;
  }

  list(): Job[] {
    return [...this.jobs.values()].map((j) => ({ ...j })).sort((a, b) => b.createdAt.localeCompare(a.createdAt));
  }

  get stats(): { total: number; active: number; queued: number } {
    const all = [...this.jobs.values()];
    return {
      total: all.length,
      active: all.filter((j) => j.status === "running").length,
      queued: all.filter((j) => j.status === "queued").length,
    };
  }

  private pump(): void {
    while (this.active < this.concurrency && this.queue.length > 0) {
      const id = this.queue.shift();
      if (!id) break;
      const job = this.jobs.get(id);
      if (!job || job.status !== "queued") continue;
      void this.run(job);
    }
  }

  private async run(job: Job): Promise<void> {
    const runner = this.runners.get(job.kind);
    if (!runner) {
      job.status = "failed";
      job.error = "runner vanished";
      this.event(job, "failed");
      return;
    }
    this.active += 1;
    job.status = "running";
    job.startedAt = new Date().toISOString();
    this.event(job, "started");
    try {
      job.result = await runner(job.payload, job, (p) => {
        job.progress = Math.max(0, Math.min(100, p));
      });
      job.progress = 100;
      job.status = "done";
    } catch (err) {
      job.status = "failed";
      job.error = err instanceof Error ? err.message : String(err);
    } finally {
      job.finishedAt = new Date().toISOString();
      this.active -= 1;
      this.event(job, job.status);
      this.pump();
    }
  }

  private event(job: Job, phase: string): void {
    const line =
      JSON.stringify({ ts: new Date().toISOString(), phase, id: job.id, kind: job.kind, status: job.status, progress: job.progress }) +
      "\n";
    try {
      appendFileSync(join(this.eventDir, "jobs.jsonl"), line);
    } catch {
      return;
    }
  }
}

File: tokyox/modules/jobs/task_manager.py
from __future__ import annotations
import asyncio
import json
import os
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Callable, Awaitable

from ..memory.twin import TwinMemory


@dataclass
class Job:
    id: str
    kind: str
    payload: Any
    status: Literal["queued", "running", "done", "failed", "cancelled"] = "queued"
    progress: int = 0
    result: Any = None
    error: str | None = None
    created_at: str = field(default_factory=lambda: time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()))
    started_at: str | None = None
    finished_at: str | None = None


RunnerFn = Callable[[Any, Job, Callable[[int], None]], Awaitable[Any]]


class TaskManager:
    def __init__(self, event_dir: str, concurrency: int = 2):
        self._event_dir = event_dir
        self._concurrency = concurrency
        self._jobs: dict[str, Job] = {}
        self._queue: list[str] = []
        self._runners: dict[str, RunnerFn] = {}
        self._active = 0
        self._pump_task: asyncio.Task | None = None
        os.makedirs(event_dir, exist_ok=True)
        self._events_file = os.path.join(event_dir, "jobs.jsonl")

    def register(self, kind: str, runner: RunnerFn) -> None:
        self._runners[kind] = runner

    def submit(self, kind: str, payload: Any) -> Job:
        if kind not in self._runners:
            raise ValueError(f"no runner registered for kind '{kind}'")
        job = Job(
            id=f"job_{uuid.uuid4().hex[:10]}",
            kind=kind,
            payload=payload,
        )
        self._jobs[job.id] = job
        self._queue.append(job.id)
        self._event(job, "queued")
        self._pump()
        return job

    def cancel(self, id_: str) -> bool:
        job = self._jobs.get(id_)
        if not job:
            return False
        if job.status == "queued":
            self._queue = [j for j in self._queue if j != id_]
            job.status = "cancelled"
            job.finished_at = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
            self._event(job, "cancelled")
            return True
        return False

    def get(self, id_: str) -> Job | None:
        return self._jobs.get(id_)

    def list(self) -> list[Job]:
        return sorted(self._jobs.values(), key=lambda j: j.created_at, reverse=True)

    @property
    def stats(self) -> dict[str, int]:
        all_jobs = list(self._jobs.values())
        return {
            "total": len(all_jobs),
            "active": sum(1 for j in all_jobs if j.status == "running"),
            "queued": sum(1 for j in all_jobs if j.status == "queued"),
        }

    def _pump(self) -> None:
        if self._pump_task and not self._pump_task.done():
            return
        self._pump_task = asyncio.create_task(self._pump_loop())

    async def _pump_loop(self) -> None:
        while self._active < self._concurrency and self._queue:
            id_ = self._queue.pop(0)
            job = self._jobs.get(id_)
            if not job or job.status != "queued":
                continue
            await self._run(job)

    async def _run(self, job: Job) -> None:
        runner = self._runners.get(job.kind)
        if not runner:
            job.status = "failed"
            job.error = "runner vanished"
            self._event(job, "failed")
            return
        self._active += 1
        job.status = "running"
        job.started_at = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        self._event(job, "started")
        try:
            job.result = await runner(job.payload, job, lambda p: setattr(job, "progress", max(0, min(100, p))))
            job.progress = 100
            job.status = "done"
        except Exception as err:
            job.status = "failed"
            job.error = str(err)
        finally:
            job.finished_at = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
            self._active -= 1
            self._event(job, job.status)
            self._pump()

    def _event(self, job: Job, phase: str) -> None:
        line = json.dumps({
            "ts": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "phase": phase,
            "id": job.id,
            "kind": job.kind,
            "status": job.status,
            "progress": job.progress,
        }) + "\n"
        try:
            with open(self._events_file, "a") as f:
                f.write(line)
        except Exception:
            pass